field/src/coset.rs

```diff
@@ -0,0 +1,222 @@
+    /// the shift untouched. Note that new coset is contained in the original one.
+    /// Returns `None` if `log_scale_factor` is greater than factor is larger
+    /// than size of `self.log_size()`.
+    pub fn shrink_coset(&self, log_scale_factor: usize) -> Option<Self> {
+        self.log_size
+            .checked_sub(log_scale_factor)
+            .map(|new_log_size| TwoAdicMultiplicativeCoset {
+                shift: self.shift,
+                log_size: new_log_size,
+            })
+    }
+
+    /// Returns the coset `self^(2^log_scale_factor)` (i. e. with shift and
+    /// subgroup generator equal to the `2^log_scale_factor`-th power of the
+    /// original ones). Returns `None` if `log_scale_factor` is greater than
+    /// factor is larger than size of `self.log_size()`.
+    pub fn exp_power_of_2(&self, log_scale_factor: usize) -> Option<Self> {
+        self.shrink_coset(log_scale_factor).map(|mut coset| {
+            coset.shift = self.shift.exp_power_of_2(log_scale_factor);
+            coset
+        })
+    }
+
+    /// Returns a new coset where the shift has been set to `shift` times the
+    /// original shift.
+    pub fn shift_by(&self, shift: F) -> TwoAdicMultiplicativeCoset<F> {
+        TwoAdicMultiplicativeCoset {
+            shift: self.shift * shift,
+            log_size: self.log_size,
+        }
+    }
+
+    /// Returns a new coset where the shift has been set to `shift`
+    pub fn set_shift(&self, shift: F) -> TwoAdicMultiplicativeCoset<F> {
+        TwoAdicMultiplicativeCoset {
+            shift,
+            log_size: self.log_size,
+        }
+    }
+
+    /// Checks if the given field element is in the coset
+    pub fn contains(&self, element: F) -> bool {
+        // Note that, in a finite field F (this is not true of a general finite
+        // commutative ring), there is exactly one subgroup of |F^*| of order n
+        // for each divisor n of |F| - 1, and its elements e are uniquely
+        // caracterised by the condition e^n = 1.
+
+        // We check (shift^{-1} * element)^(2^log_size) = 1, which is equivalent
+        // to checking shift^(2^log_size) = element^(2^log_size) - this avoids
+        // inversion at the cost of a few squarings. The loop terminates early
+        // if possible.
+        let (mut shift, mut element) = (self.shift, element);
+
+        for _ in 0..self.log_size {
+            if element == shift {
+                return true;
+            }
+            element = element.square();
+            shift = shift.square();
+        }
+
+        element == shift
+    }
+
+    /// Returns the element `shift * generator^index`, which is the `index %
+    /// self.size()`-th element of `self` (and, in particular, the `index`-th
+    /// element of `self` whenever `index` < self.size()).
+    #[inline]
+    pub fn element(&mut self, index: usize) -> F {
+        self.shift * self.generator_exp(index & ((1 << self.log_size) - 1))
+    }
```

Comment:
Given the change to `generator_exp` you don't need the `&` here.
```suggestion
    #[inline]
    pub fn element(&mut self, index: usize) -> F {
        self.shift * self.generator_exp(index)
    }
```